### PR TITLE
feat(http-client): Forward Requests from response interceptors

### PR DESCRIPTION
**Pull request by davismj**
If a response interceptor returns a new Request, the fetch client resolves a fetch call to that Request.

### CODE REVIEW

**Comment by davismj** [comment]
@AshleyGrant Here is a PR that implements the feature originally presumed by @CasiOo. You can forward a `Request` from a response interceptor and it will rerun the fetch. This makes refetching trivial (see https://github.com/aurelia/fetch-client/pull/105/files#diff-69d1ba4b313b5f4fc7ca97fff2bf4ae8R273).

The upside is that this adds some powerful options including but not limited to retry. It also makes adding retry as a first class citizen as simple as a 5 line interceptor and the configuration code fromhttps://github.com/aurelia/fetch-client/pull/104. 

The downside is that response interceptors probably expect responses. I doubt this is going to be a breaking change for anyone, as if a `Response` is returned the developer is likely responsible. However, it might be prudent to add a distinct implementation for [`processResponse`](https://github.com/aurelia/fetch-client/blob/master/src/http-client.js#L208) that aborts immediately in the event that a `Response` is encountered.

Feedback welcome. @eriklieben @EisenbergEffect

File: src/http-client.js [hunk 1]
```diff
@@ -104,24 +104,64 @@ export class HttpClient {
...
         } else {
           throw new Error(`An invalid result was returned by the interceptor chain. Expected a Request or Response instance, but got [${result}]`);
         }
 
-        return request.then(_request => processResponse(response, this.interceptors, _request));
+        return processResponse(response, this.interceptors, request);
+      })
+      .then(result => {
+        if (Request.prototype.isPrototypeOf(result)) {
+          debugger;
```
Comment (AshleyGrant) [inline src/http-client.js]: debugger statement left in.

**Comment by davismj** [comment]
> Finally, this code doesn't actually allow us to fully implement the RetryInterceptor with the beforeRetry callback "in the box", as the interceptor still does not have access to the HttpClient instance, which is key to the use case of allowing the developer to possibly update the configuration of the HttpClient before doing the retry.

I've added the `HttpClient` as a new argument to the interceptors.

File: src/http-client.js (hunk 1 above)
Comment (davismj) [reply src/http-client.js]: image

**Pull request merged**